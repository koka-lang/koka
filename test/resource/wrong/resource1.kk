// --------------------------------------------------------
// resources
// test with safe resources using rank-2 types to prevent
// escaping the correct scope
// --------------------------------------------------------

effect heap<s::S> { }

fun heap( action : forall<s> () -> <heap<s>|e> a ) : e a {
  handle<heap<_>>(action) { }
}

effect resource ref<s::S,a::V> in heap<s> {
  fun get() : a
  fun set( value : a ) : ()
}

val new-ref = handler resource(s) {
  get() -> resume(s,s)
  set(x) -> resume((),x)
}

fun foo() {
  using heap
  use r1 = new-ref(36)  // () -> <heap<s>|e> ref<s,int>
  r1
}

fun main() {
  using heap
  val r = foo()
  r.get.println
}
