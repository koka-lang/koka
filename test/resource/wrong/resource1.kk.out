test/resource/wrong/resource1.kk(24, 3): error: type is not polymorph enough
 context : using heap
 term : using heap
 use r1 = new-ref(36)
 inferred type : forall<s1> () -> <heap<s1>,div> ref<s1,int>
 is less polymorph as: forall<s1> () -> <heap<s1>,div|_e> ref<$s,int>